forbid enum values that are out of range

fbenkstein opened the pull request:
Enum values that are out of range can lead to generated C++ code that does
not compile.  Also forbid boolean enums.

aardappel commented on src/idl_parser.cpp:
```diff
@@ -1566,6 +1590,12 @@ CheckedError Parser::ParseEnum(bool is_union, EnumDef **dest) {
       if (Is('=')) {
         NEXT();
         ev.value = StringToInt(attribute_.c_str());
+        // TODO: values between INT64_MAX and UINT64_MAX get silently truncated
```
We still want these kinds of values to be representable, so the correct behavior is to store them bitwise in a signed number

aardappel commented:
Yes, @fbenkstein can you see what happens in @vglavnyy's PR? Which PR should go in first?

fbenkstein commented:
@vglavnyy should go in first.  I'll rebase my stuff on top of his.

aardappel commented:
https://github.com/google/flatbuffers/pull/4948 has been merged.

fbenkstein commented:
@vglavnyy @aardappel I've rebased and found that #4948 didn't address the out of range enums.  I've adjusted the error message to be consistent.  It will also print the valid interval as that's pretty useful.  There's still the question of very large ulong enum values (between INT64_MAX and UINT64_MAX).  I haven't found an issue referencing these, so maybe we should create one?

vglavnyy commented on tests/test.cpp:
```diff
@@ -1227,7 +1227,7 @@ void ErrorTest() {
   TestError("enum X:byte { Y, Y }", "value already");
   TestError("enum X:byte { Y=2, Z=1 }", "ascending");
-  TestError("union X { Y = 256 }", "must fit");
+  TestError("union X { Y = 256 }", "enum value does not fit");
```
I don't understand this test.
This declaration `union Any { Monster, Weapon, Pickup }` is correct.
But `union X { Y = 256 }` is incorrect. Here we refer to undefined type `256` not number.

vglavnyy replied:
AInUnionAgain:A = 10 is not equal to AInUnionAgain = 10 from a sematic point of view.
In any case, this is a minor issue, I think this will be resolved automatically.
Extra check if (ev.value < 0 || ev.value >= 256) will be redurant.

fbenkstein commented on src/idl_parser.cpp:
```diff
@@ -1649,6 +1673,16 @@ CheckedError Parser::ParseEnum(bool is_union, EnumDef **dest) {
       if (Is('=')) {
         NEXT();
         ECHECK(atot(attribute_.c_str(), *this, &ev.value));
+        ev.value = StringToInt(attribute_.c_str());
```
Yes, you are right.  Using `ParseSingleValue` makes the code much shorter now.  I've kept denying boolean enums even though you could define `enum X:bool { A = true, B = false }` which could make sense because it doesn't currently forbid `enum X:bool { A = 100 }` which doesn't make sense in my opinion.  It would also need special handling in the code generators, I guess.

vglavnyy replied:
If @aardappel agree with bool-base enums it will be very simple:
- comment or remove one line of code
- update grammar documentation

 Boolean values as a base for enum was disabled before refactoring, I only save this behavior.

aardappel replied:
I don't see a use for boolean enums, since `bool` in FlatBuffers always takes one byte. So you get the same as using `ubyte` except now your enum is not extensible. I'd suggest to not allow `bool`.

vglavnyy replied:
@aardappel 
Is following declaration acceptable?
`enum X:ubyte { A = true, B = false }`

Or simply: Can boolean constants **true** and **false** are used for implicit initialization (1, 0) of an integer-type values?
Now **true** or **false** can be applied only to a bool-type value.

aardappel replied:
I suppose we could allow these as integers in any context (by moving them to be parsed as integers in the lexical analysis), though I don't see the need / use case for this.

fbenkstein commented:
@vglavnyy @aardappel Is there something still to be done here or can this go in?

vglavnyy commented:
Probably I was wrong.
Folowing code isn't redurant:
```
      if (is_union) {
        if (ev.value < 0 || ev.value >= 256)
          return Error("union enum value must fit in a ubyte");
      }
```
If someone create an union with 256 or more tables:
```
union tables_0_256 {
 t0, .. , t255, t256
}
```
the parser accepts it as valid declaration.
This is the normal case without `=` assignment.
Generated values for this union aren't checked to be in range.

It will be better to check that all values are in range regardless direct assignment `=` and value of `is_union` flag.

fbenkstein commented:
@vglavnyy Good catch.  I will try to add tests for that as well.

fbenkstein commented:
@vglavnyy @aardappel Please have another look.

vglavnyy commented:
@fbenkstein
Why not use a 'duck' approach?
- Cast an implicit assigned int64 value to a string;
- Check that '=' case passed and modify the string if needed;
- Check string with `ParseSingleValue(check_now=true)`, next apply `atot<int64>`.

With implicit assignment (without `=`) we have three extra steps:
1) assigned ev.value to string with NumToString from `utils.h`;
2) String to a number by `ParseSingleValue::atot`;
3) String to a number by `atot<int64>`.

With explicit assignment using `=` we have:
2) String to a number by `ParseSingleValue::atot`;
3) String to a number by `atot<int64>`.

Steps (2) and (3) are common.
The code is ugly but short (minimal code duplication) and covers all cases.
What do you think?

Another way: to extract `check_now` code from `ParseSingleValue::atot` to standalone static or templated function and try to reuse it.

fbenkstein commented:
@vglavnyy 
> Another way: Extract `check_now` code from `ParseSingleValue::atot` to standalone static or templated function and try to reuse it.

I like that approach but I'd rather do that later rather than in this PR.

aardappel commented:
Ok, later changes can be in another PR, thanks!
